### ISSUE TITLE
Performance Bottleneck

### ISSUE THREAD

**Issue by giangianoulas** [eclipse-birt/birt#1633]
Hello

After running a huge report (pdf) on a customer witha very large dataset, using a custom plain file driver, we get OutOfMemmory.
While trying to figure out this one i got a java flight recording that showed a performance issue
![image](https://github.com/eclipse-birt/birt/assets/21333067/8bbddd3a-a067-49de-abcc-ebe00373b8a1)

looking at 
```org.eclipse.birt.data.engine.impl.CancelManager```

```java
	public CancelManager() {
		cancellableList = new ArrayList<>();
	}

	/**
	 *
	 * @param cancellable
	 */
	public void register(ICancellable cancellable) {
		synchronized (cancellableList) {
			cancellableList.add(cancellable);
		}
	}

	/**
	 *
	 * @param cancellable
	 */
	public void deregister(ICancellable cancellable) {
		synchronized (cancellableList) {
			cancellableList.remove(cancellable);
		}
	}
```

It apperas to be a candidate for LinkedList as it is not accessed too regularly
but elements are getting added and removed quite often.

**Comment by hvbtup** [comment]
Hmm, I don't think changing the type to a `LinkedList` instead of `ArrayList` would help. In both cases removing elements takes _O(n)_.
Shouldn't it be a `HashMap` or something like that instead?

If you are profiling anyway, it would also be interesting to see the average element count of the structure.

Maybe the size is more of a problem. Looking at the code, it seems to me that registering/deregistering only takes place in the methods of `DataSourceQuery` class.

Maybe there is some kind of _memory leak_ like forgetting to call `deregister` somehow or delaying the call too long? That would cause the structure to grow and grow, if a big report uses several thousands or millions of queries. (Well, with a classical remote RDBMS, such a report would be a candidate for refactoring anyway, because it'd run forever and slow the DB down)

Because otherwise I cannot understand why a few (or a few dozens) queries which are open at the same time can cause the structure to grow to a significant size _n_.

**Comment by speckyspooky** [comment]
Yes, I confirm the idea of Henning. I have also some reports which use many queries in combination with loops and with large row count (1Mio.) but I have never seen this problem.

Do you got the problem with the latest BIRT version 4.15?

**Comment by giangianoulas** [comment]
There was indeed a memory leak. In ```DataSourceQuery#execute``` elements were added but never removed.
Also there was a unnecessary big syncronized block.

I agree that this report needs a redesign, but it is not that easy to be done.
I am using Birt 4.12.0 but it seems to be ancient problem

**Comment by hvbtup** [comment]
Can you provide a PR?

**Comment by giangianoulas** [comment]
Ok all done
equals and hashCode was not necessary so i removed them.

Sadly i can produce this amount of data because it is quite complex and i don’t have the business knowledge for this.

I will come back to this and comment when i will have something at hand

**Comment by merks** [comment]
@giangianoulas 

FYI, a build with your changes is available here:

https://download.eclipse.org/birt/updates/nightly/latest

**Comment by giangianoulas** [comment]
I did some minor testing. 
With the same report as the one that initially had the problem but with a dataset at least **one order of magintute** smaller.
I did run the old build and the newer build 10 times each, measured the execution time (Yes i know this is wildly inaccurate and almsot invalid) but i saw 3% faster execution time.
It is what i actually excpected, this problem only kicked in with large datasets and the larger the worst it got, but still it is a measurable improvement
